core/net/ip/tcpip.c

```diff
@@ -526,222 +522,268 @@ tcpip_input(void)
-      }
-#else /* UIP_ND6_SEND_NA */
-      PRINTF("tcpip_ipv6_output: neighbor not in cache\n");
-      uip_len = 0;
-      return;  
-#endif /* UIP_ND6_SEND_NA */
+  nbr = uip_ds6_nbr_lookup(nexthop);
+  if(nbr == NULL) {
+    if(send_nd6_ns(nexthop)) {
+      PRINTF("tcpip_ipv6_output: failed to add neighbor to cache\n");
+      goto exit;
     } else {
+      goto send_packet;
+    }
+  }
+
 #if UIP_ND6_SEND_NA
-      if(nbr->state == NBR_INCOMPLETE) {
-        PRINTF("tcpip_ipv6_output: nbr cache entry incomplete\n");
-#if UIP_CONF_IPV6_QUEUE_PKT
-        /* Copy outgoing pkt in the queuing buffer for later transmit and set
-           the destination nbr to nbr. */
-        if(uip_packetqueue_alloc(&nbr->packethandle, UIP_DS6_NBR_PACKET_LIFETIME) != NULL) {
-          memcpy(uip_packetqueue_buf(&nbr->packethandle), UIP_IP_BUF, uip_len);
-          uip_packetqueue_set_buflen(&nbr->packethandle, uip_len);
-        }
-#endif /*UIP_CONF_IPV6_QUEUE_PKT*/
-        uip_clear_buf();
-        return;
-      }
-      /* Send in parallel if we are running NUD (nbc state is either STALE,
-         DELAY, or PROBE). See RFC 4861, section 7.3.3 on node behavior. */
-      if(nbr->state == NBR_STALE) {
-        nbr->state = NBR_DELAY;
-        stimer_set(&nbr->reachable, UIP_ND6_DELAY_FIRST_PROBE_TIME);
-        nbr->nscount = 0;
-        PRINTF("tcpip_ipv6_output: nbr cache entry stale moving to delay\n");
-      }
+  if(nbr->state == NBR_INCOMPLETE) {
+    PRINTF("tcpip_ipv6_output: nbr cache entry incomplete\n");
+    queue_packet(nbr);
+    goto exit;
+  }
+  /* Send in parallel if we are running NUD (nbc state is either STALE,
+     DELAY, or PROBE). See RFC 4861, section 7.3.3 on node behavior. */
+  if(nbr->state == NBR_STALE) {
+    nbr->state = NBR_DELAY;
+    stimer_set(&nbr->reachable, UIP_ND6_DELAY_FIRST_PROBE_TIME);
+    nbr->nscount = 0;
+    PRINTF("tcpip_ipv6_output: nbr cache entry stale moving to delay\n");
+  }
 #endif /* UIP_ND6_SEND_NA */
 
-      tcpip_output(uip_ds6_nbr_get_ll(nbr));
+send_packet:
+  if(nbr) {
+    linkaddr = uip_ds6_nbr_get_ll(nbr);
+  } else {
+    linkaddr = NULL;
+  }
 
-#if UIP_CONF_IPV6_QUEUE_PKT
-      /*
-       * Send the queued packets from here, may not be 100% perfect though.
-       * This happens in a few cases, for example when instead of receiving a
-       * NA after sendiong a NS, you receive a NS with SLLAO: the entry moves
-       * to STALE, and you must both send a NA and the queued packet.
-       */
-      if(uip_packetqueue_buflen(&nbr->packethandle) != 0) {
-        uip_len = uip_packetqueue_buflen(&nbr->packethandle);
-        memcpy(UIP_IP_BUF, uip_packetqueue_buf(&nbr->packethandle), uip_len);
-        uip_packetqueue_free(&nbr->packethandle);
-        tcpip_output(uip_ds6_nbr_get_ll(nbr));
-      }
-#endif /*UIP_CONF_IPV6_QUEUE_PKT*/
+  tcpip_output(linkaddr);
 
-      uip_clear_buf();
-      return;
-    }
+  if(nbr) {
```

Comment:
I would prefer having this block conditioned by `UIP_CONF_IPV6_QUEUE_PKT` (rather than the body of `send_queued`).
Would be clearer IMO